class.jetpack.php

```diff
@@ -738,11 +732,65 @@ function( $callable ) {
 		add_filter( 'jetpack_token_request_body', array( __CLASS__, 'filter_token_request_body' ) );
 	}
 
+	/**
+	 * Runs on plugins_loaded but with a higher priority to make sure we're initializing
+	 * before other hooks.
+	 *
+	 * @action plugins_loaded
+	 * @priority 5
+	 */
+	public function early_initalization() {
+		/*
+		 * Enable enhanced handling of previewing sites in Calypso
+		 */
+		if ( self::is_active() ) {
+			require_once JETPACK__PLUGIN_DIR . '_inc/lib/class.jetpack-iframe-embed.php';
+			add_action( 'init', array( 'Jetpack_Iframe_Embed', 'init' ), 9, 0 );
+			require_once JETPACK__PLUGIN_DIR . '_inc/lib/class.jetpack-keyring-service-helper.php';
+			add_action( 'init', array( 'Jetpack_Keyring_Service_Helper', 'init' ), 9, 0 );
+		}
+
+		Sync_Main::init();
+
+		// Check for WooCommerce support.
+		Sync_Actions::initialize_woocommerce();
```

Comment:
I found it odd that we initialize these here instead of in the `sync` package. 

Should we call an action here and hook to it from the sync package instead?

Reply:
Yes, this is one of the things I wanted to discuss as well. Since `Sync` is now a package, it shouldn't initalize anything on its own, right? The application that initalizes Sync should have some means of control over what Sync has to initialize.

So this is just a first attempt, maybe we need a separate method, sort of like WordPress themes declare support for different functionality, we can make something like `Sync_Main::enable_woo_support()`?

Reply:
>  Since Sync is now a package, it shouldn't initialize anything on its own, right?

I'm certain about that, but it makes sense. This is a decision we must make in the design of packages (not only the sync one), and we should be consistent with it. It makes sense from packaging point of view, but if we follow that pattern, we should do a couple more things:

* Document the initialization properly.
* Make sure to explain that initialization is handled by the package consumer.
* Try to keep initialization as simple and as straightforward as possible (not 20 actions, but ideally 1 or 2).
* Be consistent and follow this pattern for all packages.

Let's think about how these packages will be consumed. Eventually, let's imagine Woo using connection and sync packages - do they have to initialize it all, or would they expect it would come initialized by default? Both would work, of course, but let's think about what makes most sense.

Reply:
> maybe we need a separate method, sort of like WordPress themes declare support for different functionality, we can make something like Sync_Main::enable_woo_support()?

Sounds good to me, and pretty usable!

Reply:
if a dev is going to use a package I guess they want it initialized, maybe we should add filters to turn it off in special cases instead?

Reply:
in any case, I agree with @tyxla that is supper odd to have the woo sync initialization here.